### ISSUE TITLE
redirect to form action pathname

### ISSUE THREAD

**Issue by david-plugge** [sveltejs/kit#8936]
### Describe the bug

```html
<form action="/signin" use:enhance>...</form>
```
Lets assume this form lives inside `src/routes/+layout.svelte` and mirrors the form inside `src/routes/signin/+page.svelte`. With js disabled, we get redirected to `/signin` and the formdata is correctly populated. If js is enabled, we stay at the route we are currently at. This is inconsistent, the browser should go to `/signin`.

### Reproduction

https://github.com/david-plugge-reproductions/sveltekit-actions-redirect

### System Info

```Shell
System:
    OS: Linux 5.15 Ubuntu 22.04.1 LTS 22.04.1 LTS (Jammy Jellyfish)
    CPU: (12) x64 AMD Ryzen 5 5600X 6-Core Processor
    Memory: 11.56 GB / 15.57 GB
    Container: Yes
    Shell: 5.8.1 - /usr/bin/zsh
Binaries:
    Node: 18.13.0 - /mnt/wslg/runtime-dir/fnm_multishells/6383_1675825091098/bin/node
    npm: 8.19.3 - /mnt/wslg/runtime-dir/fnm_multishells/6383_1675825091098/bin/npm
npmPackages:
    @sveltejs/adapter-auto: ^2.0.0 => 2.0.0
    @sveltejs/kit: ^1.5.0 => 1.5.0
    svelte: ^3.54.0 => 3.55.1
    vite: ^4.0.0 => 4.1.1
```

### Severity

annoyance

**Comment by dummdidumm** [comment]
This works as designed. It allows you to place forms in arbitrary places while staying there when JS is enabled. You could have `<form>` with a newsletter signup inside your layout so it's visible at the bottom of each page. You don't want to interrupt the user if they submit their email by redirecting them to the signup page.

**Comment by david-plugge** [comment]
While this is true, it doesn´t work with js disabled. This sounds weird to me as the issue about layout actions (https://github.com/sveltejs/kit/issues/8462) was closed. So we are able to partially use *globally* available form actions with js but not without?

**Comment by dummdidumm** [comment]
I don't know what you mean by "doesn't work without". You can reference a form associated with a page (and therefore form actions in `+page.server.js`) from anywhere. With `use:enhance`, you stay on wherever you are, without JS you'll be redirected to the page the form originally lives on.

**Comment by david-plugge** [comment]
I´m thinking of modals for example. If i create a modal with a form inside the root layout, it works fine on every page with `use:enhance`. But without the client will be redirected to the actions pathname. If i create a form with an action `/newsletter?/subscribe` i´d expect to get redirected to that exact url (which is the native case). It seems like the current behaviour is trying to mimik the functionality that was asked in #8462 but i think that forms, that don´t live in the page next to the action should be treated different.

So in short i expect `use:enhance` to completely follow the native form submit behaviour.
I´m aware this is probably bad ux but that´s how forms work. To compensate this we would need something like layout actions that are available in multiple the subtree.

**Comment by dummdidumm** [comment]
It would also be a breaking change. I feel like the majority of use cases / people would prefer to stay on the page they submitted the form on. You can also adjust `use:enhance` to function as you want it.
```html
<form use:enhance={() => { return async ({ update, action }) => { await goto(action); update(); } }} ..>
```

**Comment by david-plugge** [comment]
> I feel like the majority of use cases / people would prefer to stay on the page they submitted the form on.

I totally agree and that´s why i think #8462 should get revived and `use:enhance` should follow native behaviour. I´ll create a repo to explain myself a little better.

**Comment by david-plugge** [comment]
I´ve updated the repro

https://github.com/david-plugge-reproductions/sveltekit-actions-redirect/blob/main/src/routes/%2Blayout.svelte
https://github.com/david-plugge-reproductions/sveltekit-actions-redirect/blob/main/src/routes/%2Blayout.server.ts

**Comment by dummdidumm** [comment]
This all sounds more related to #8462 , I'm not sure if that was implemented that this means we should follow native behavior and redirect. I still don't fully understand why you would want this behavior in the first place, what's the motivation?

**Comment by david-plugge** [comment]
You´re right, my issue is definitely more related to #8462, should have posted over there instead.
Just to clarify, i don´t want to get redirected to `/newsletter` when submitting the form on lets say `/`. I just don´t understand why we don´t have a way to submit that form on `/` and render that exact route again (not using `use:enhance`), with the form data populated. This is a rather big limitation (again, imagine different types of modals signin, newsletter, ...). What i currently have to do is adding a `+page.svelte` for every single action. But again, the newsletter form does not need a seperate page when it can be added to the layout footer for example.

I just wonder why we should be able to use actions out of the actual pathname´s scope (`/newsletter?/subscribe` in `src/routes/+layout.svelte`) using `use:enhance` but have no way to do this with full progressive enhancement support (render the current page).

In cases js actually fails the client **will** get redirected to the actions pathname which isn´t good ux at all. So i´d say i don´t want this behaviour, i even want to completely get rid of it.

**Comment by dummdidumm** [comment]
There are a couple of things speaking against this:
- typings -> what would this mean for them? They are not well-formed anymore. Could be a gotcha / "I know what I'm doing" thing though
- it's impossible to reliably know where the user was originally coming from when POSTing to a page. There's the referrer header [but it's not reliable](https://stackoverflow.com/questions/3096928/php-when-submitting-an-html-form-does-the-http-referer-get-set).

**Comment by david-plugge** [comment]
> * typings -> what would this mean for them? They are not well-formed anymore. Could be a gotcha / "I know what I'm doing" thing though

Yeah this is where it gets complicated. I´m a typescript guy but this feature would add so much value and i´d be fine manually adding types. But i´m sure there´s a way this can be done in a good way.

> * it's impossible to reliably know where the user was originally coming from when POSTing to a page. There's the referrer header [but it's not reliable](https://stackoverflow.com/questions/3096928/php-when-submitting-an-html-form-does-the-http-referer-get-set).

Why would we need this information? We would only care about where the action points to.

**Comment by dummdidumm** [comment]
> In cases js actually fails the client will get redirected to the actions pathname which isn´t good ux at all. So I would say I don't want that behavior, in fact I want to get rid of it altogether.

I interpreted this as "I don't want to get redirected to the page the action belongs to".

I'm now very confused about what is actually the feature request 😄

**Comment by david-plugge** [comment]
I am very sorry, hard to find good words to describe what I am talking about.
My *issue* consists of two parts

- mirror native form behaviour
- add ability to submit forms from anywhere and render the correct page that is set in the forms `action` attribute

If you have some time, I could try to explain it on Discord.

**Comment by dummdidumm** [comment]
Not sure this would change much to take this to Discord. If it helps, you can write down your request in german.

- re "mirror native form behavior" -> this one I understood. I was confused later on because it sounded like you want the opposite (full-page-reload behavior stays on the page your POST from)
- add ability to submit forms from anywhere -> sounds like your first point. That small code snippet I posted would make this possible. I don't see the value in making this the default behavior for progressively enhanced forms, so I would need a motivating example for that

**Comment by david-plugge** [comment]
First of, i love sveltekit actions. It´s a beautiful solution to the rather difficult task of implementing good forms.

But there´s this one limiting factor: an action always needs to have a `+page.svelte` file next to it. This page is rendered on submit when not using `use:enhance` or when js is not available (it failed to load due to bad connection, blocked, ...).

Example:

_src/routes/newsletter/+page.server.ts_

```ts
export const actions = {
    async subscribe() {
        ...
    }
} satisfies Actions;
```

_src/routes/+layout.svelte_

```html
<script lang="ts">
	import type { LayoutActionData } from './$types';

	export let form: LayoutActionData;
</script>

<main>
	<slot />
</main>

<!-- the current way -->
<footer>
	<form action="/newsletter?/subscribe">...</form>
</footer>

<!-- idea -->
<footer>
	<form action="?/newsletter=subscribe">...</form>
</footer>
```

Imagine a client is currently at `/account`. He has a bad internet connection and js failed to load properly.
Now he fills in the form fields and submits the data.

### The current case

When the form is submitted the server receives a request:

- url: https://your-domain.com/newsletter?/subscribe

  we set a path in the form action so the client will post to `/newsletter`.

- action:
  - file: _src/routes/newsletter/+page.server.ts_
  - name: subscribe

SvelteKit will render `src/routes/newsletter/+page.svelte` with the populated action data.
The client gets a completely different page than before but with the same form.

This is not the case when `use:enhance` is added to the form. We get different results for the same request.
(thats why i created this issue)

Here is what i´m thinking: Is this bad UX? Should the client always return to the page where he was previously filling out the form?

### The alternative

When the form is submitted the server receives a request:

- url: https://your-domain.com/account?/newsletter=subscribe

  we didn´t set a path in the form action so the client will just post to the current page.

- action:
  - file: _src/routes/newsletter/+page.server.ts_
  - name: subscribe

We can execute the action and then render the page that matches the actions pathname. In this case it´s `/account`, because that´s the page we basically asked for.

"I want to render this exact page, after running that specific action".

### Additional thoughts

Thinking about this, it would probably be better to group all these _globally_ available actions into a folder, that way it´s easy to add types and keep them seperated from _page bound_ actions.
